fish-rust/src/bitset.rs

```diff
@@ -0,0 +1,275 @@
+use num_traits::bounds::UpperBounded;
+use num_traits::{AsPrimitive, FromPrimitive, Num};
+use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, Not, Shl, Shr};
+
+pub struct BitSet<T>(T);
+
+impl<
+        T: BitAnd<T, Output = T>
+            + BitOr<T, Output = T>
+            + BitAndAssign<T>
+            + BitOrAssign<T>
+            + Shl<usize, Output = T>
+            + Shr<usize, Output = T>
+            + Not<Output = T>
+            + AsPrimitive<u8>
+            + FromPrimitive
+            + UpperBounded
+            + Num
+            + AsI64
+            + AsUsize,
+    > BitSet<T>
+where
+    i64: AsPrimitive<T>,
+{
+    pub fn set(&mut self, i: usize) {
+        self.0 |= T::one() << i;
+    }
+
+    pub fn unset(&mut self, i: usize) {
+        self.0 &= !(T::one() << i);
+    }
+
+    pub fn toggle(&mut self, i: usize, v: bool) {
+        // if v { self.set(i) } else { self.unset(i) }
+        let v = T::from_u8(v as u8).unwrap();
+        let mask = T::one() << i;
+        let bit: T = ((-v.as_i64()).wrapping_shr(63)).as_();
```

Comment:
Why is this shift necessary? If I got this right, `v` is either `0_T` or `1_T` here, which is then converted to either `0_i64` or `-1_i64` - shifting either of those right by 63 is a no-op. Wouldn't `(-i64::from(v)).as_()` (without L35 above) be the same thing?

Reply:
You're right - great catch. I was originally trying another approach and confused myself into making this more complicated than it needed to be.